python/pyarrow/tests/test_io.py

```diff
@@ -160,6 +188,24 @@ def test_inmemory_write_after_closed():
     with pytest.raises(IOError):
         f.write(b'not ok')
 
+def test_buffer_protocol_ref_counting():
+    import gc
+
+    val = b'data'
+    buf = io.buffer_from_bytes(val)
+    as_bytes = bytes(buf)
+
+    def add_temporary_reference(buf):
+        m = bytearray(buf)
+        del(buf)
```

Comment:
This doesn't do what you want (leaving only a last "internal" reference to the memory) because the outer `buf` reference is still valid. I would suggest

```
def make_buffer(bytes_obj):
    return bytearray(io.buffer_from_bytes(bytes_obj))
buf = make_buffer(b'foo')
```

Here, `buf` is the only owner of the data (I think!)